File: java_version/linked_lists/linked_list_find/linked_list_find.java
package linked_lists.linked_list_find;

// linked list find
// Write a function, linkedListFind, that takes in a pointer to the head of a linked list and a target value.
//  The function should return a boolean indicating whether or not the linked list contains the target.
class Node<T> {
    T val;
    Node<T> next;

    public Node(T val) {
        this.val = val;
        this.next = null;
    }
}

public class linked_list_find {

    public static <T> boolean linkedListFind(Node<T> head, T target) {
        // todo
    

        return false;
    }

    public static <T> boolean linkedListFindRecur(Node<T> head, T target) {
        // todo
            return true;
    }

    public static <T> void _linkedListFindRecur(Node<T> head, T target) {
        // todo
        return;
    }

    public static void main(String[] args) {
        Node<String> a = new Node<>("a");
        Node<String> b = new Node<>("b");
        Node<String> c = new Node<>("c");
        Node<String> d = new Node<>("d");
        a.next = b;
        b.next = c;
        c.next = d;
        System.out.println(linkedListFind(a, "c")); // true
        System.out.println(linkedListFindRecur(a, "c")); // true

        System.out.println(linkedListFind(a, "d")); // true
        System.out.println(linkedListFindRecur(a, "d")); // true

        System.out.println(linkedListFind(a, "q")); // false
        System.out.println(linkedListFindRecur(a, "q")); // false

        Node<String> node1 = new Node<>("jason");
        Node<String> node2 = new Node<>("leneli");
        node1.next = node2;
        System.out.println(linkedListFind(node1, "jason")); // true
        System.out.println(linkedListFindRecur(node1, "jason")); // true

        Node<Integer> node3 = new Node<>(42);
        System.out.println(linkedListFind(node3, 42)); // true
        System.out.println(linkedListFindRecur(node3, 42)); // true

        System.out.println(linkedListFind(node3, 100)); // false
        System.out.println(linkedListFindRecur(node3, 100)); // false

    }
}

File: javascript_version/Linked_lists/linked_list_find/linked_list_find.js
// linked list find
// Write a function, linkedListFind, that takes in a pointer to the head of a linked list and a target value.
//  The function should return a boolean indicating whether or not the linked list contains the target.


class Node {
    constructor (value) {
        this.value = value;
        this.next = null;
    }
}


const linkedListFind = (head, target) => {

}

const linkedListFindRecur = (head, target) => {

}

const _linkedListFindRecur = (curr, target) => {


}

console.log("iterative method");

const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");

a.next = b;
b.next = c;
c.next = d;

// a -> b -> c -> d

console.log(linkedListFind(a, "c")); // 1 (true)

// const a = new Node("a");
// const b = new Node("b");
// const c = new Node("c");
// const d = new Node("d");

// a.next = b;
// b.next = c;
// c.next = d;

// a -> b -> c -> d

console.log(linkedListFind(a, "d")); // 1 (true)
console.log(linkedListFind(a, "q")); // 0 (false)


console.log("recursion method");
console.log(linkedListFindRecur(a, "c")); // 1 (true)
console.log(linkedListFindRecur(a, "d")); // 1 (true)
console.log(linkedListFindRecur(a, "q")); // 0 (false)

const node1 = new Node("jason");
const node2 = new Node("leneli");

node1.next = node2;

// jason -> leneli
console.log("iterative method");

console.log(linkedListFind(node1, "jason")); // 1 (true)
console.log("recursion method");
console.log(linkedListFindRecur(node1, "jason")); // 1 (true)

const node3 = new Node(42);

// 42
console.log("iterative method");
console.log(linkedListFind(node3, 42)); // true
console.log("recursion method");
console.log(linkedListFindRecur(node3, 42)); // true



const node10 = new Node(42);

// 42
console.log("iterative method");

console.log(linkedListFind(node10, 100)); // false

console.log("recursion method");
console.log(linkedListFindRecur(node10, 100)); // false

//recursive method 

